File: go-backend-interview-prep/02_data_structures/04_maps/main.py
# Python equivalent of Maps (compare with main.go)
# Go map -> Python dict

def main():
    # --- Create with literal ---
    ages = {"alice": 30, "bob": 25}
    print("ages:", ages)

    # --- Create empty ---
    scores = {}
    scores["math"] = 95
    scores["english"] = 88
    print("scores:", scores)

    # --- Read: KeyError for missing keys (unlike Go zero value) ---
    print("\n--- Read ---")
    print("alice:", ages["alice"])
    # ages["nobody"]  # would raise KeyError (Go returns zero value)

    # --- .get() is Python's comma-ok equivalent ---
    print("\n--- .get() (like comma-ok) ---")
    val = ages.get("alice", 0)  # default if missing
    print(f"alice: val={val}")

    val = ages.get("nobody", 0)
    print(f"nobody: val={val}")

    # --- Check existence with `in` ---
    print("alice exists:", "alice" in ages)
    print("nobody exists:", "nobody" in ages)

    # --- Write ---
    ages["charlie"] = 35
    print("\nafter add charlie:", ages)

    # --- Update ---
    ages["alice"] = 31
    print("after update alice:", ages)

    # --- Delete ---
    del ages["bob"]
    print("after delete bob:", ages)

    # Delete missing: use pop with default to avoid KeyError
    ages.pop("nobody", None)

    # --- Length ---
    print("length:", len(ages))

    # --- Iteration order (insertion order since Python 3.7) ---
    print("\n--- Iteration (insertion order) ---")
    colors = {"red": "#ff0000", "green": "#00ff00", "blue": "#0000ff"}
    for k, v in colors.items():
        print(f"  {k} -> {v}")
    print("(Python preserves insertion order; Go does NOT)")

    # --- No nil dict trap in Python ---
    # None-type dict just doesn't have methods.
    # But: d = None; d["key"] = 1  -> AttributeError

    # --- Word frequency (same pattern) ---
    print("\n--- Word frequency ---")
    words = ["go", "is", "great", "go", "is", "fast", "go"]
    freq = {}
    for w in words:
        freq[w] = freq.get(w, 0) + 1
    for word, count in freq.items():
        print(f"  {word}: {count}")

    # --- Set ---
    print("\n--- Set ---")
    items = ["a", "b", "a", "c", "b"]
    unique = set(items)
    print("unique count:", len(unique))


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/02_data_structures/04_maps/main.go
package main

import "fmt"

func main() {
	// --- Create with literal ---
	ages := map[string]int{
		"alice": 30,
		"bob":   25,
	}
	fmt.Println("ages:", ages)

	// --- Create with make ---
	scores := make(map[string]int)
	scores["math"] = 95
	scores["english"] = 88
	fmt.Println("scores:", scores)

	// --- Read: zero value for missing keys ---
	fmt.Println("\n--- Read ---")
	fmt.Println("alice:", ages["alice"])     // 30
	fmt.Println("missing:", ages["nobody"])  // 0 (zero value, no panic)

	// --- Comma-ok idiom ---
	fmt.Println("\n--- Comma-ok ---")
	val, ok := ages["alice"]
	fmt.Printf("alice: val=%d exists=%t\n", val, ok)

	val, ok = ages["nobody"]
	fmt.Printf("nobody: val=%d exists=%t\n", val, ok)

	// --- Write ---
	ages["charlie"] = 35
	fmt.Println("\nafter add charlie:", ages)

	// --- Update ---
	ages["alice"] = 31
	fmt.Println("after update alice:", ages)

	// --- Delete ---
	delete(ages, "bob")
	fmt.Println("after delete bob:", ages)

	// Delete a missing key: no panic, no-op
	delete(ages, "nobody")

	// --- Length ---
	fmt.Println("length:", len(ages))

	// --- Iteration order is RANDOM ---
	fmt.Println("\n--- Iteration (random order) ---")
	colors := map[string]string{
		"red":   "#ff0000",
		"green": "#00ff00",
		"blue":  "#0000ff",
		"white": "#ffffff",
	}
	for k, v := range colors {
		fmt.Printf("  %s -> %s\n", k, v)
	}
	fmt.Println("(run again -- order may differ)")

	// --- Nil map trap ---
	fmt.Println("\n--- Nil map ---")
	var nilMap map[string]int
	fmt.Println("nil map read:", nilMap["key"]) // 0, no panic
	fmt.Println("nil map len:", len(nilMap))     // 0
	// nilMap["key"] = 1  // PANIC: assignment to entry in nil map

	// --- Count word frequency (common pattern) ---
	fmt.Println("\n--- Word frequency ---")
	words := []string{"go", "is", "great", "go", "is", "fast", "go"}
	freq := make(map[string]int)
	for _, w := range words {
		freq[w]++ // zero value of int is 0, so this works cleanly
	}
	for word, count := range freq {
		fmt.Printf("  %s: %d\n", word, count)
	}

	// --- Set pattern (map[T]bool or map[T]struct{}) ---
	fmt.Println("\n--- Set pattern ---")
	seen := make(map[string]struct{})
	items := []string{"a", "b", "a", "c", "b"}
	for _, item := range items {
		seen[item] = struct{}{}
	}
	fmt.Println("unique count:", len(seen))
}
